### PR TITLE
[#128857207] Add unit tests for slack bot

### PR DESCRIPTION
**Pull request by nivivon**
I started creating a factories.js but couldn't get it to work properly. Figured based on the nature of these tests that it's not necessary just yet. The tests and app.js should be fine though.

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+var _ = require('lodash'); // jshint ignore:line
+var gruntOc = require('grunt-opencare');
+
```
Comment (leeopencare) [inline Gruntfile.js:5]: Can we clean up this grunt file?

File: Gruntfile.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    'request',
+    'authCommon'
+  ];
+
+  var srcGlobals = [
+    '_',
+    'Q',
+    'S',
+    'moment',
+    'sails',
```
Comment (leeopencare) [inline Gruntfile.js]: Alot of these are not needed.

File: Gruntfile.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  gruntOc.tasks.release(grunt);
+
+  var srcFiles = ['**/*.js', '!node_modules/**/*.js', '!test/**/*.js', '!.tmp/**/*.js'];
+  var testFiles = ['test/*.js'];
+
+  var testGlobals = [
+    'describe',
+    'it',
+    'before',
+    'after',
```
Comment (leeopencare) [inline Gruntfile.js:18]: Same here.

File: Gruntfile.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        USERNAME: 'TEST_USERNAME',
+        PASSWORD: 'TEST_PASSWORD',
+        SLACK_WEBHOOK_URI: 'b',
+        SLACK_RELEASE_WEBHOOK_URI: 'c',
+        SLACK_API_TOKEN: 'd'
+      }
+    });
+
+    grunt.registerTask('run', ['nodemon:dev']);
+    grunt.registerTask('tunnel', ['ngrok:api:tls:1337', 'run']);
```
Comment (leeopencare) [inline Gruntfile.js]: remove.

File: app.js [hunk 5]
```diff
@@ -80,56 +66,66 @@ app.post('/', function(req, res) {
...
     } else { // Feedback given
-      resType = ResType.API;
-
-      if (payload.comment.user.login in githubToSlack) {
-        username = githubToSlack[payload.issue.user.login];
-        text = payload.comment.user.login + ' commented on your PR in <' + payload.repository.html_url + '|' + payload.repository.name + '>: ' + payload.comment.body + '\n<' + payload.issue.html_url + '|' + payload.issue.title + '>';
+      resType = common.ResType.API;
+      if (!(payload.issue.user.login in common.Github.ToSlack)) {
+        return res.json(400, {
+          error: payload.issue.user.login + " is not defined"
```
Comment (leeopencare) [inline app.js]: Use `'`

File: app.js [hunk 6]
```diff
@@ -80,56 +66,66 @@ app.post('/', function(req, res) {
...
+  } else if (resType == common.ResType.API) {
+    options = {
       text: text,
       channel: '@' + username,
       username: slackBotUsername,
       icon_url: slackBotIconURL
-    }, handleResponse(res));
+    };
+    if (process.env.NODE_ENV === 'test') {
+      return res.json(200, _.merge(options, {
```
Comment (leeopencare) [inline app.js]: I am grateful 🎧  that you wrote these tests but I think responding with a 200 only in test mode does not go full circle, I think we should use nock to emulate the slack API to make sure it goes all the way.

**Comment by vadim-zverugo** [comment]
🚀

File: test/slackbot.spec.js [hunk 7]
```diff
@@ -14,6 +12,18 @@ describe('slackbot', function() {
   var url = 'http://' + process.env.USERNAME + ':' + process.env.PASSWORD + '@localhost:8080/';
   var emoji = '\uD800\uDC00';
 
+  var nock;
+  before(function() {
+    nock = require('nock');
+    // nock.recorder.rec();
```
Comment (leeopencare) [inline test/slackbot.spec.js]: Remove

File: test/slackbot.spec.js [hunk 8]
```diff
@@ -24,71 +34,86 @@ describe('slackbot', function() {
...
     it('should send a message on opening a pull request', function(done) {
+      nock('https://hooks.slack.com:443')
+        .post('/services/T024HT77N/B035LH8PN/IzKEPzLeV2rTNWway46FAwdu', {
+          "response_type": "ephemeral",
+          "text": ":rocket: nivivon opened a pull request in <example.com|example>\n*<example.com|test pr>*",
+          "username": "github",
+          "link_names": 0,
+          "icon_emoji": ""
+        })
+        .reply(200, "ok");
```
Comment (leeopencare) [inline test/slackbot.spec.js]: Could be on previous line, remove `"` where possible.

**Comment by leeopencare** [comment]
👯

**Comment by nivivon** [comment]
Merged this in pr2slack-opencare. This repo is public so no longer using it.

**Pull request closed without merging**